src/infra/core/src/transform/transform_helpers.rs

```diff
@@ -0,0 +1,309 @@
+    let output_chain = target.dataset.as_metadata_chain();
+
+    // TODO: externalize
+    let block_ref = BlockRef::Head;
+    let head = output_chain.resolve_ref(&block_ref).await.int_err()?;
+
+    // TODO: PERF: Search for source, vocab, and data schema result in full scan
+    let (source, schema, set_vocab, prev_query) = {
+        // TODO: Support transform evolution
+        let mut set_transform_visitor = SearchSetTransformVisitor::new();
+        let mut set_vocab_visitor = SearchSetVocabVisitor::new();
+        let mut set_data_schema_visitor = SearchSetDataSchemaVisitor::new();
+        let mut execute_transform_visitor = SearchExecuteTransformVisitor::new();
+
+        target
+            .dataset
+            .as_metadata_chain()
+            .accept_by_hash(
+                &mut [
+                    &mut set_transform_visitor,
+                    &mut set_vocab_visitor,
+                    &mut set_data_schema_visitor,
+                    &mut execute_transform_visitor,
+                ],
+                &head,
+            )
+            .await
+            .int_err()?;
+
+        (
+            set_transform_visitor.into_event(),
+            set_data_schema_visitor
+                .into_event()
+                .as_ref()
+                .map(SetDataSchema::schema_as_arrow)
+                .transpose() // Option<Result<SchemaRef, E>> -> Result<Option<SchemaRef>, E>
+                .int_err()?,
+            set_vocab_visitor.into_event(),
+            execute_transform_visitor.into_event(),
+        )
+    };
+
+    let Some(source) = source else {
+        return Err(TransformNotDefinedError {}.into());
+    };
+    tracing::debug!(?source, "Transforming using source");
+
+    // Prepare inputs
+    use itertools::Itertools;
+    let input_states: Vec<(TransformInput, Option<ExecuteTransformInput>)> =
+        if let Some(query) = &prev_query {
+            source
+                .inputs
+                .iter()
+                .cloned()
+                .zip_eq(query.query_inputs.iter().cloned().map(Some))
+                .collect()
+        } else {
+            source.inputs.iter().map(|i| (i.clone(), None)).collect()
+        };
+
+    // Build preliminary transform request
+    Ok(TransformPreliminaryRequestExt {
+        operation_id: get_random_name(None, 10),
+        dataset_handle: target.handle.clone(),
+        block_ref,
+        head,
+        transform: source.transform,
+        system_time,
+        schema,
+        prev_offset: prev_query.as_ref().and_then(ExecuteTransform::last_offset),
+        vocab: set_vocab.unwrap_or_default().into(),
+        input_states,
+        prev_checkpoint: prev_query.and_then(|q| q.new_checkpoint.map(|c| c.physical_hash)),
+    })
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub(crate) async fn get_transform_input_from_query_input(
```

Comment:
They are only called from elaboration service because of compacting feature, which forces the partial replanning activity over already resolved datasets.

I would rather rethink the need in that feature.